src/ddmd/cond.d

```diff
@@ -64,6 +67,227 @@ extern (C++) abstract class Condition : RootObject
+        fd.fbody = s;
+        auto fe = new FuncExp(loc, fd);
+        auto ce = new CallExp(loc,fe,new Expressions());
+        return ce;
+    }
+
+    private extern(D) Statement createForeach(Loc loc, Parameters* parameters, Statement s)
+    {
+        if (aggrfe)
+        {
+            return new ForeachStatement(loc, aggrfe.op, parameters, aggrfe.aggr.syntaxCopy(), s, loc);
+        }
+        else
+        {
+            assert(!!rangefe && parameters.dim == 1);
+            return new ForeachRangeStatement(loc, rangefe.op, (*parameters)[0], rangefe.lwr.syntaxCopy(), rangefe.upr.syntaxCopy(), s, loc);
+        }
+    }
+
+    private extern(D) TypeStruct createTupleType(Loc loc, Expressions* e, Scope* sc)
+    {   // TODO: move to druntime?
+        auto sid = Identifier.generateId("Tuple");
+        auto sdecl = new StructDeclaration(loc, sid);
+        sdecl.storage_class |= STCstatic;
+        sdecl.members = new Dsymbols();
+        auto fid = Identifier.idPool(StaticForeach_tupleFieldName.ptr, StaticForeach_tupleFieldName.length);
+        auto ty = new TypeTypeof(loc, new TupleExp(loc, e));
+        sdecl.members.push(new VarDeclaration(loc, ty, fid, null, 0));
+        auto r = cast(TypeStruct)sdecl.type;
+        r.vtinfo = TypeInfoStructDeclaration.create(r); // prevent typeinfo from going to object file
+        return r;
+    }
+
+    private extern(D) Expression createTuple(Loc loc, TypeStruct type, Expressions* e)
+    {   // TODO: move to druntime?
+        return new CallExp(loc, new TypeExp(loc, type), e);
+    }
+
+    private void lowerNonArrayAggregate(Scope* sc)
+    {   // TODO: move to druntime?
+        auto nvars = aggrfe ? aggrfe.parameters.dim : 1;
+        auto aloc = aggrfe ? aggrfe.aggr.loc : rangefe.lwr.loc;
+        Parameters*[3] pparams = [new Parameters(),new Parameters(),new Parameters()];
+        foreach(i;0..nvars)
+        {
+            foreach(params;pparams)
+            {
+                auto p = aggrfe ? (*aggrfe.parameters)[i] : rangefe.prm;
+                params.push(new Parameter(p.storageClass, p.type, p.ident, null));
+            }
+        }
+        Expression[2] res;
+        TypeStruct tplty = null;
+        if (nvars == 1)
+        {
+            foreach(i;0..2)
+            {
+                res[i] = new IdentifierExp(aloc, (*pparams[i])[0].ident);
+            }
+        }
+        else
+        {
+            foreach(i;0..2)
+            {
+                auto e = new Expressions();
+                foreach (j;0..pparams[0].dim)
+                {
+                    auto p = (*pparams[i])[j];
+                    e.push(new IdentifierExp(aloc, p.ident));
+                }
+                if (!tplty)
+                {
+                    tplty = createTupleType(aloc, e, sc);
+                }
+                res[i] = createTuple(aloc, tplty, e);
+            }
+            needExpansion = true;
+        }
+        auto s1 = new Statements();
```

Comment:
This code is obviously building some sort of AST, but I have no idea what. Please express in a comment what is being generated.